activerecord/test/cases/associations/has_many_through_associations_test.rb

```diff
@@ -953,6 +953,13 @@ def test_can_update_through_association
     end
   end
 
+  def test_has_many_through_polymorphic_with_scope_should_work_properly
+    post = TaggedPost.create!(title: "Tagged", body: "Post")
+    tag = post.tags.create!(name: "Tag")
+    assert_equal [tag], TaggedPost.preload(:tags).last.tags
+    assert_equal [tag], TaggedPost.eager_load(:tags).last.tags
```

Comment:
Actually the issue also happens with eager loading. So I'v fixed it in https://github.com/rails/rails/pull/29154/commits/6c70ce89c8e1698af2735898cc7c423b3b8fc3f7.